osu.Server.Queues.ScorePump/ImportHighScores.cs

```diff
@@ -125,9 +148,47 @@ void queueNextBatch()
                 }
             }
 
+            Console.WriteLine($"Finished in {(DateTimeOffset.Now - start).TotalSeconds} seconds.");
             return 0;
         }
 
+        private void checkSlaveLatency(MySqlConnection db)
+        {
+            long currentTimestamp = DateTimeOffset.Now.ToUnixTimeSeconds();
+
+            if (currentTimestamp - lastLatencyCheckTimestamp < seconds_between_latency_checks)
+                return;
+
+            lastLatencyCheckTimestamp = currentTimestamp;
+
+            // This latency is best-effort, and randomly queried from available hosts (with rough precedence of the importance of the host).
+            // When we detect a high latency value, a recovery period should be introduced where we are pretty sure that we're back in a good
+            // state before resuming operations.
+            int latency = db.QueryFirst<int>("SELECT `count` FROM `osu_counts` WHERE NAME = 'slave_latency'");
+
+            if (latency > maximum_slave_latency_seconds)
+            {
+                Console.WriteLine($"Current slave latency of {latency} seconds exceeded maximum of {maximum_slave_latency_seconds} seconds.");
+                Console.WriteLine($"Sleeping for {latency} seconds to allow catch-up.");
+
+                Thread.Sleep(Math.Max(seconds_between_latency_checks * 2, latency * 1000));
+
+                // greatly reduce processing rate to allow for recovery.
+                scoresPerQuery = Math.Max(safe_minimum_scores_per_query, scoresPerQuery / 2);
+            }
+
+            if (latency > 0)
```

Comment:
So basically the only case where processing rate is allowed to be increased is when the slave latency is 0 seconds?

Reply:
yep. it's usually always the case that it's 0 (or low enough that it floors to zero).